### PR TITLE
Preserve order in left join for cudf-polars

### PR DESCRIPTION
**Pull request by wence-**
## Description

Unlike all other joins, polars provides an ordering guarantee for left joins. By default libcudf does not, so we need to order the gather maps in this case.

While here, because it requires another hard-coding of `int32` for something that should be `size_type`, expose `type_to_id` in cython and plumb it through.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/_lib/pylibcudf/libcudf/utilities/type_dispatcher.pxd [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+
+from libcpp cimport bool
+from libcpp.vector cimport vector
```
Comment (brandon-b-miller) [inline python/cudf/cudf/_lib/pylibcudf/libcudf/utilities/type_dispatcher.pxd:4]: are these imports needed?

Comment (wence-) [reply python/cudf/cudf/_lib/pylibcudf/libcudf/utilities/type_dispatcher.pxd:4]: Oh, no, they are not.

**Comment by wence-** [comment]
Fixed merge conflicts, ready for another look.

**Review by vyasr — APPROVED** [review]

File: python/cudf/cudf/_lib/pylibcudf/types.pyx [hunk 2]
```diff
@@ -67,3 +68,7 @@ cdef class DataType:
         cdef DataType ret = DataType.__new__(DataType, type_id.EMPTY)
         ret.c_obj = dt
         return ret
+
+
+SIZE_TYPE = DataType(type_to_id[size_type]())
```
Comment (vyasr) [inline python/cudf/cudf/_lib/pylibcudf/types.pyx:73]: This seems like something that could probably be used much more broadly than it is in this PR. Not in scope, but something we might want to come back to and clean up.

Comment (wence-) [reply python/cudf/cudf/_lib/pylibcudf/types.pyx:73]: Yeah, I was just introducing it here really.

**Pull request merged**